_posts/2019-10-23-kafka-authentication-using-oauth-2.0.md

```diff
@@ -0,0 +1,538 @@
+---
+layout: post
+title:  "Kafka authentication using OAuth 2.0"
+date: 2019-10-23
+author: marko_strukelj
+---
+
+In Strimzi 0.14 we have added an additional authentication option to the standard set supported by Kafka brokers.
+Your Kafka clients can now use OAuth 2.0 token-based authentication when establishing a session to a Kafka broker.
+With this kind of authentication Kafka clients and brokers talk to a central OAuth 2.0 compliant authorization server. 
+Clients use the authorization server to obtain access tokens, or are configured with access tokens issued by the server.
+Brokers communicate with authorization server to validate the tokens presented by clients, thus confirming their identities.
+In this model, the administrator doesn't create users in Kafka.
+Rather, users are typically managed through identity providers exposed over LDAP.
+Strimzi client components like Kafka Bridge, Kafka Connect, and Kafka Mirror Maker can be configured with OAuth 2.0 authentication through custom resource definitions.
+
+
+<!--more-->
+
+## Authentication in Kafka
+
+When using Apache Kafka you have several options to implement authentication. 
+You can use TLS with client certificates to establish a secure channel and authenticate a client at the same time.
+Another option is to separately configure TLS for connection encryption and integrity, but use a specific SASL protocol for authentication.
+
+There are several SASL based protocols you can use.
+One option is to use SASL PLAINTEXT to authenticate with username and password.
+Another option is to use SASL SCRAM which is also username and password based, but rather than storing clear text passwords it stores salted hashes.
```

Comment:
AFAIK the passwords are as hashes not only stored but also not transfered over the network which is worth mentioning.

Reply:
Do you have a suggestion how to word this?
Looking at [documentation](https://kafka.apache.org/documentation/#security_sasl_scram) it doesn't exactly jump out if passwords are communicated in the clear between Kafka client and broker, but since client is configured with SCRAM mechanism I'd say the SCRAM hash is communicated, not username, and password - I may be wrong. Regardless, using SCRAM does not mean you are safe without also using TLS between Kafka client and broker but also between broker and ZooKeeper.

Reply:
I think that with SCRAM it is even a bit more complicated than that. It does some challenge / response communication, it doesn't transfer the hash it self. Anyway, I think the current form is sufficient.